drivers/sdhc/rcar_mmc.c

```diff
@@ -0,0 +1,1472 @@
+ *
+ * @param dev The pointer to an MMC device structure
+ *
+ * @retval none
+ */
+static void rcar_mmc_init_host_props(const struct device *dev)
+{
+	struct mmc_rcar_data *data = dev->data;
+	const struct mmc_rcar_cfg *cfg = dev->config;
+	struct sdhc_host_props *props = &data->props;
+	struct sdhc_host_caps *host_caps = &props->host_caps;
+
+	memset(props, 0, sizeof(*props));
+
+	/* Note: init only properties that are used for mmc/sdhc */
+
+	props->f_max = cfg->max_frequency;
+	/* note: actually, it's possible to get lower frequency
+	 *       if we use divider from cpg too
+	 */
+	props->f_min = (cfg->max_frequency >> 9);
+
+	props->power_delay = 100; /* ms */
+
+	props->is_spi = 0;
+
+	/* TODO: add support of 4/8 bits bus */
+	host_caps->bus_8_bit_support = 0;
+	host_caps->bus_4_bit_support = 0;
+
+	/* TODO: add support to driver */
+	host_caps->high_spd_support = 0;
+
+	host_caps->vol_330_support = 1;
+	host_caps->vol_300_support = 0;
+	host_caps->vol_180_support = 1;
+
+	host_caps->sdr50_support = 1;
+	host_caps->sdr104_support = cfg->mmc_sdr104_support;
+	host_caps->ddr50_support = 1;
+	host_caps->uhs_2_support = cfg->uhs_support;
+	host_caps->hs200_support = cfg->mmc_hs200_1_8v;
+	host_caps->hs400_support = cfg->mmc_hs400_1_8v;
+}
+
+/**
+ * @brief Reset sampling clock controller registers
+ *
+ * @param cfg The Renesas MMC driver configuration
+ *
+ * @retval none
+ */
+static void rcar_mmc_disable_scc(const struct mmc_rcar_cfg *cfg)
+{
+	uint32_t reg;
+
+	/* just to be to be sure that the SD clock is disabled */
+	 rcar_mmc_enable_clock(cfg, false);
+
+	/* Reset SCC registers, need to disable and enable clock
+	 * before and after reset
+	 */
+
+	/* Disable SCC sampling clock */
+	reg = rcar_mmc_read_reg32(cfg, RENESAS_SDHI_SCC_CKSEL);
+	reg &= ~RENESAS_SDHI_SCC_CKSEL_DTSEL;
+	rcar_mmc_write_reg32(cfg, RENESAS_SDHI_SCC_CKSEL, reg);
+
+	/* disable hs400 mode & data output timing */
+	reg = rcar_mmc_read_reg32(cfg, RENESAS_SDHI_SCC_TMPPORT2);
+	reg &= ~(RENESAS_SDHI_SCC_TMPPORT2_HS400EN |
+			 RENESAS_SDHI_SCC_TMPPORT2_HS400OSEL);
+	rcar_mmc_write_reg32(cfg, RENESAS_SDHI_SCC_TMPPORT2, reg);
+
+	rcar_mmc_enable_clock(cfg, true);
+
+	/* disable SCC sampling clock position correction */
+	reg = rcar_mmc_read_reg32(cfg, RENESAS_SDHI_SCC_RVSCNTL);
+	reg &= ~RENESAS_SDHI_SCC_RVSCNTL_RVSEN;
+	rcar_mmc_write_reg32(cfg, RENESAS_SDHI_SCC_RVSCNTL, reg);
+}
+
+/**
+ * @brief Initialize and configure the Renesas MMC controller registers
+ *
+ * @param dev The pointer to an MMC device structure
+ *
+ * @retval none
+ */
+static void rcar_mmc_init_controller_regs(const struct device *dev)
+{
+	uint32_t reg;
+	struct mmc_rcar_data *data = dev->data;
+	const struct mmc_rcar_cfg *cfg = dev->config;
+	struct sdhc_io ios = {0};
+
+	rcar_mmc_reset(dev);
+
+	if (rcar_mmc_card_busy(dev)) {
+		LOG_WRN("%s: device busy, some of registers can't be initilized",
+			dev->name);
+	}
+
+	/* Disable SD clock (SD_CLK) output
```

Comment:
Probably here should be empty line at the start of block comment, please use script/checkpatch.pl before commit.